client/DESIGN.md

```diff
@@ -0,0 +1,292 @@
+
+Additionally, every "table:key" key in the view also stores two booleans called "updated" and "deleted" alongside its sorted map.  The "updated" boolean is False if the Transaction has only read the value of that key and True if the Transaction has written an update to that key's value.  The "deleted" boolean is True if and only if the user has called the delete method on its key during this Transaction.
+
+Upon commit, we search through the map keys for updated values (updated==True) and deleted values (deleted==True) and construct a set of op\_lists from the most recent value (sorted by (LONG\_MAX - read\_timestamps)) in each updated key's sorted map.  Note the op\_list is a parameter to Cache's write method, and each table needs its own op\_list.
+
+## Transaction Public Methods
+
+Note that each of the methods below first attempts to contact the cache to complete its operation.  If the cache does not contain the necessary information, the cache makes the HTTP request to the HTTP server.  (The Transaction does not directly make HTTP requests.)
+
+### read
+    read(
+        table: string, 
+        key: string, 
+        read_timestamp=currentTime, 
+        max_age=None, 
+        no_cache=False) returns value (JSON) or None
+
+The read method first attempts to read the specified key from the view, then from cache, and finally from the DB server.  If read finds the requested value, it stores it in the view; otherwise, it returns None and stores the key had no value (in Python, None) for that read_timestamp satisfying the given parameters.
+
+### write
+    write(
+        table: string, 
+        key: string, 
+        value: JSON) returns None (always succeeds)
+
+The write method updates the Transaction view to reflect the specified write.  (They key immediately has its "updated" status set to True.)  Note that writes never immediately fail (are never immediately rejected) because they are only reflected in the view until we commit the Transaction.
+
+Write adds the new value to its sorted map with the current time as both the read and value timestamp.
+
+* **Question (7)**: Should we set both the read and value timestamps of new entries to the current time?
+
+### delete
+    delete(
+        table: string, 
+        key: string) returns None
+
+If the key exists in the Transaction view, set its "deleted" boolean to True.  Otherwise, add the key to the view with "deleted" set to True and the value None.
+
+### commit
+    commit() returns None, throws StaleException
+
+Initialize an operation dictionary D mapping (table identifier strings) to (operation lists).  Iterate through keys in the view.  For each "table:key" key in the view, consider its sorted map value, along with its "updated" and "deleted" flags.  
```

Comment:
We are translating all entries in the view from `table : key -> (op : value)` to an array of `[(table, key, op, value)]`---mostly, the ops `hold` and `delete` don't need a value. We could just skip the conversion and hand the map to Cache.write. It can iterate the map directly to write out the HTTP request for the batch.